main.tf

```diff
@@ -0,0 +1,163 @@
+// TO DO: - Docs to note the required Cloud Function configs (public ingress - auth required)
+//        - all resources must be in the same project
+//        - functions in OpenAPI spec must also be in TF vars
+//        - security config must match what's in example yaml file for API auth to work 
+
+// TO DO: Create an example
+
+// locals
+locals {
+  configmd5 = md5(file("open-api.yaml")) # this generates an md5 hash so that the api config name is unique whenever the config changes
+  apis      = ["apikeys.googleapis.com", "apigateway.googleapis.com"]
+}
+
+// enable required APIs for this module
+resource "google_project_service" "apigw" {
+  for_each = toset(local.apis)
+  project  = var.project_id
+  service  = each.key
+
+  timeouts {
+    create = "30m"
+    update = "40m"
+  }
+
+  disable_on_destroy = false
+}
+
+// create random suffix for resources to ensure uniqueness
+resource "random_id" "default" {
+  byte_length = 2
+}
+
+// create google cloud service account
+resource "google_service_account" "default" {
+  project      = var.project_id
+  account_id   = "api-gw-service-account-${random_id.default.hex}"
+  display_name = "Service Account created by TF API GW module"
+}
+
+// get cloud functions
+data "google_cloudfunctions2_function" "default" {
+  count    = length(var.cloud_functions)
+  project  = var.project_id
+  name     = var.cloud_functions[count.index].name
+  location = var.cloud_functions[count.index].location
+}
+
+// add "roles/run.invoker" to each cloud function
+resource "google_cloud_run_service_iam_member" "default" {
+  count    = length(data.google_cloudfunctions2_function.default)
+  location = data.google_cloudfunctions2_function.default[count.index].location
+  project  = var.project_id
+  service  = data.google_cloudfunctions2_function.default[count.index].name
+  role     = "roles/run.invoker"
+  member   = "serviceAccount:${google_service_account.default.email}"
+}
+
+// add Cloud API Gateway Service Agent role to service account
+resource "google_project_iam_member" "default" {
+  project = var.project_id
+  role    = "roles/apigateway.serviceAgent"
+  member  = "serviceAccount:${google_service_account.default.email}"
+}
+
+// Create API GW + Configs 
+resource "google_api_gateway_api" "api_gw" {
+  provider     = google-beta
+  project      = var.project_id
+  api_id       = "api-gw-${random_id.default.hex}"
+  display_name = "API GW config created by TF module"
+}
+
+resource "google_api_gateway_api_config" "api_gw" {
+  provider      = google-beta
+  project       = var.project_id
+  api           = google_api_gateway_api.api_gw.api_id
+  api_config_id = "config-${local.configmd5}"
+  display_name  = "API GW config created by TF module"
```

Comment:
Would be good to pass this in as a variable

Reply:
Ez, will do this.